aten/src/ATen/native/cuda/Loops.cuh

```diff
@@ -286,6 +287,25 @@ static void launch_kernel(int64_t N, const func_t& f, array_t data) {
 
 } // namespace modern
 
+template<typename func_t, int nargs=function_traits<func_t>::arity>
+struct needs_dynamic_casting {
+  static bool check(TensorIterator& iter) {
+    using traits = function_traits<func_t>;
+    if (iter.dtype(nargs) != c10::impl::CPPTypeToScalarType<typename traits::template arg<nargs - 1>::type>::value) {
```

Comment:
What happens here if one of the args is a scalar (eg a Python float)?

Reply:
I think it will be wrapped as a CPU scalar tensor then get removed in `gpu_kernel_with_scalars`.

Reply:
```C++
template <typename func_t>
void gpu_kernel_with_scalars(TensorIterator& iter, const func_t& f) {
  ASSERT_HOST_DEVICE_LAMBDA(func_t);
  TORCH_INTERNAL_ASSERT(iter.ntensors() == 3);

  using traits = function_traits<func_t>;
  static_assert(
      traits::arity == 2,
      "gpu_kernel_with_scalars only supports two input arguments");

  if (iter.is_cpu_scalar(1)) {
    using arg1_t = typename traits::template arg<0>::type;
    using arg2_t = typename traits::template arg<1>::type;
    auto a = iter.scalar_value<arg1_t>(1);
    iter.remove_operand(1);
    gpu_kernel(iter, [=]GPU_LAMBDA(arg2_t b) {
      return f(a, b);
    });
  } else if (iter.is_cpu_scalar(2)) {
    using arg1_t = typename traits::template arg<0>::type;
    using arg2_t = typename traits::template arg<1>::type;
    auto b = iter.scalar_value<arg2_t>(2);
    iter.remove_operand(2);
    gpu_kernel(iter, [=]GPU_LAMBDA(arg1_t a) {
      return f(a, b);
    });
  } else {
    gpu_kernel(iter, f);
  }
}
```

Reply:
```C++
Tensor mul(const Tensor& self, Scalar other) {
  return native::mul(self, wrapped_scalar_tensor(other));
}
```

Reply:
It's also worth noting that it will be wrapped either in Double or Long Scalar, so I hope all subsequent casts are correct :-)

Reply:
```C++
// These are still needed because we don't have C++ conversions from number
// types (int, float, etc.) to Tensor (only to Scalar). They're not exposed
// to Python.

static Tensor wrapped_scalar_tensor(Scalar scalar) {
  auto tensor = scalar_to_tensor(scalar);
  tensor.unsafeGetTensorImpl()->set_wrapped_number(true);
  return tensor;
}
```

Reply:
I think you're right.  I also think that's a problem.  When that scalar argument gets stripped out, gpu_kernel_with_scalars forwards a lambda that only has 1 argument, not 2.  Now your templates are comparing iter.dtype(nargs) for an iter with 2 arguments against the argument types of a lambda with only 1 argument.

Reply:
@mcarilli The iter will also have only one argument. Because of `iter.remove_operand(1);` or `iter.remove_operand(2);`

Reply:
Ok, `remove_operand` was the missing ingredient I didn't spot.  Neat!